### PR TITLE
Fixes `Display` and Adds Other Formatting Traits

### PR DESCRIPTION
**Pull request by dtoniolo**
# TL;DR
Hi, thanks for accepting my previous pull request. This one fixes what I feel like is an issue with the current implementation of `Display` and adds support for the `LowerExp` and `UpperExp` formatting traits. With this PR the newtypes of `strict-num` would implement all and only the formatting traits that are avaiable on the floating point primitive types. Morover, with the exception of `Debug`, their implementation would be equivalent to the one of `f32` and `f64`.

As before, I'm open to any feedback or change proposal.

## Details
The current version of `strict-num` ignores the formatting arguments that are passed to `Display`, but I would expect to be able to format `strict-num`'s newtypes as if they were their corresponding primitive type.

Accodingly, the first of the two commits swaps the current macro-based implementation for the one from `derive_more`, which simply forwards to the `Display::fmt` method of the corresponding primitive type.

The main drawback of these changes is that they require adding a new dependency, which increases compile times. On the flip side, `derive_more` makes it easy to implement a few other traits, such as:
- `LowerExp` and `UpperExp`, that are added in the second commit.
- Conversion traits as `From<PositiveF32>` for `FiniteF32` and `f32`, which could be added in a future pull request.

Moreover, `derive_more` is fully compatible with a `#![no_std]` environment, so it doesn't break `strict-num`'s support for embedded development. Finally, `derive_more` is a completely private dependency[^1], so it wouldn't cause any compatibility problems for downstream crates.

[^1]: By this I mean that nothing from `derive_more` would be exposed to downstream crates, so there's no risk of conflict with other dependencies that require incompatibile versions of `derive_more`.

## A Note About Testing
Testing these traits in a `#![no_std]` environment has been tricky because it's impossibile to use the `format!` macro. I used a solution (adapted from StackOverflow) that relies on fixed-size, stack allocated strings. It would be better to rely on the [`heapless`](https://docs.rs/heapless/0.8.0/heapless/) crate, but one of its dependencies requires Rust 1.60+. I don't think that it's worth increasing the MSRV for such a minimal usage of a test utility, but if this PR is accepted we might want to open an issue to keep track of this problem, so that when the MSRV will be increased again it will be easy to remember of [`heapless`].

Edit: added note about `#![no_std]`.

[`String`]: https://docs.rs/heapless/0.8.0/heapless/struct.String.html
[`heapless`]: https://docs.rs/heapless/0.8.0/heapless/

### CODE REVIEW

**Comment by RazrFalcon** [comment]
`derive_more` is a proc-macro and no-go immediately. I avoid proc-macros as a plague. It's my own preference. 

As for the PR, I do agree that the current formatter is rudimentary and if you can improve it I will gladly accept the changes.

As for `no_std` I'm a bit confused. To implement better formatting we need `format!`, but it requires `std`? But how `derive_more` handles it then? Maybe we should simply hide implementations with `format!` behind a `cfg`?

**Comment by dtoniolo** [comment]
Hi, sorry for the late answer, I have been really busy recently. 

Regaring `format!` I explained myself badly, I'll elabore more. The `format!` macro creates a new `String`, which requires heap allocation and is thus a no-go in `#![no_std]` environments. The various formatting traits, however, don't require `String`s: they write on a string buffer which can be located anywhere - even on the stack. Moreover, these traits aren't defined in the standard library, but in [`core::fmt`](https://doc.rust-lang.org/nightly/core/fmt/index.html). `derive_more` derives them using the [`write!`] and [`format_args!`] macros from the same module.

It's therefore perfectly possible to implement and use the formatting traits in `#![no_std]` environments. Given, however, that in these environments it's not possible to use `format!`, they have to write into fixed-size, stack allocated strings that have to be created in advance. This is what I had to do for testing[^testing_note]. In particular, I wrapped all the functionaliy needed in a tailor-made struct, which is called `WriteTo` in the two commits I pushed[^written].

The logic that `WriteTo` implements is already provided by another crate, [`heapless`]. While their version is surely more robust and better written, adding this crate to the dependencies of `strict-num` would require increasing the MSRV and I'm not sure it's worth it.

Would be happy to elaborate further if something isn't clear.

### `derive_more`
I fully respect your choice of avoiding procedural macros. I think I can manage to derive the traits by using declarative macros and will attempt to do so over the next few days.

Just ot of curiosity, may I ask what motivates you to avoid procedural macros? I do not mean to raise any discussios about this topic, I'm asking just because I've always used procedural macros without asking myself too many questions, so it's quite likely that I've missed something important about the side effects they can introduce and other drawbacks they have.


[^written]: For the sake of intellectual honesty, much of the code of that struct was taken from an online discussion, as noted in its documentation.

[^testing_note]: I could also have circumvented the problem by enabling the standard library just for the tests of the formatting traits and use `format!`. I decided against this option because it would run the tests I thought it would be safer to run the tests in an environment that has the same limitations of the library itself, i.e. that is `#![no_std]`.

[`format_args!`]: https://doc.rust-lang.org/nightly/core/macro.format_args.html
[`heapless`]: https://crates.io/crates/heapless
[`write!`]: https://doc.rust-lang.org/nightly/core/macro.write.html

**Comment by RazrFalcon** [comment]
>Just ot of curiosity, may I ask what motivates you to avoid procedural macros?

It increases compilation times quite a lot (at least on a clean build) and increases the size of `cargo vendor` output.
